presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java

```diff
@@ -0,0 +1,759 @@
+        }
+
+        public PartitionedOutputInfo getInfo()
+        {
+            return new PartitionedOutputInfo(rowsAdded.get(), pagesAdded.get(), outputBuffer.getPeakMemoryUsage());
+        }
+
+        public void partitionPage(Page page)
+        {
+            // Populate positions to copy for each destination partition.
+            int positionCount = page.getPositionCount();
+
+            for (int i = 0; i < partitionBuffer.length; i++) {
+                partitionBuffer[i].resetPositions(positionCount);
+            }
+
+            Block nullBlock = nullChannel.isPresent() ? page.getBlock(nullChannel.getAsInt()) : null;
+            Page partitionFunctionArgs = getPartitionFunctionArguments(page);
+
+            for (int position = 0; position < positionCount; position++) {
+                boolean shouldReplicate = (replicatesAnyRow && !hasAnyRowBeenReplicated) ||
+                        nullBlock != null && nullBlock.isNull(position);
+
+                if (shouldReplicate) {
+                    for (int i = 0; i < partitionBuffer.length; i++) {
+                        partitionBuffer[i].addPosition(position);
+                    }
+                    hasAnyRowBeenReplicated = true;
+                }
+                else {
+                    int partition = partitionFunction.getPartition(partitionFunctionArgs, position);
+                    partitionBuffer[partition].addPosition(position);
+                }
+            }
+
+            // Decode the page just once. The decoded blocks will be fed to each PartitionBuffer object to set up BlockEncodingBuffers.
+            for (int i = 0; i < decodedBlocks.length; i++) {
+                decodedBlocks[i] = decodeBlock(flattener, page.getBlock(i));
+            }
+
+            // Copy the data to their destination partitions and flush when the buffer is full.
+            for (int i = 0; i < partitionBuffer.length; i++) {
+                partitionBuffer[i].appendData(decodedBlocks, fixedWidthRowSize, variableWidthChannels, outputBuffer);
+            }
+        }
+
+        public void flush()
+        {
+            for (int i = 0; i < partitionBuffer.length; i++) {
+                partitionBuffer[i].flush(outputBuffer);
+            }
+        }
+
+        private Page getPartitionFunctionArguments(Page page)
+        {
+            Block[] blocks = new Block[partitionChannels.size()];
+            for (int i = 0; i < blocks.length; i++) {
+                Optional<Block> partitionConstant = partitionConstants.get(i);
+                if (partitionConstant.isPresent()) {
+                    blocks[i] = new RunLengthEncodedBlock(partitionConstant.get(), page.getPositionCount());
+                }
+                else {
+                    blocks[i] = page.getBlock(partitionChannels.get(i));
+                }
+            }
+            return new Page(page.getPositionCount(), blocks);
+        }
+
+        private int getFixedWidthTypeSize(int channel)
```

Comment:
Thanks for the suggestion. Since it uses class member sourceTypes, I updated the method signature to
`private static int getFixedWidthTypeSize(Type type)`
and the caller passes in `sourceTypes.get(i)`